### PR TITLE
Add support for reading board version from I2C

### LINKED ISSUE #65: Read board version from i2c

### PR DESCRIPTION
**Pull request by rohbotics**
This allows people with newer boards to automatically get features
that are only availible in newer hardware revisions.

Closes #65

### CODE REVIEW

**Comment by rohbotics** [comment]
Just tested this on my 5.0 board, and it seems to be reading the version number correctly.

File: magni_bringup/scripts/launch_core.py [hunk 1]
```diff
@@ -68,10 +70,26 @@ def get_conf():
...
+        # The I2C chip is only present on 5.0 and newer boards
+        try:
+            i2cbus = smbus.SMBus(1)
+            i2cbus.write_byte(0x20, 0xFF)
+            time.sleep(0.2)
+            inputPortBits = i2cbus.read_byte(0x20)
+            boardRev = 49 + (15 - (inputPortBits & 0x0F))
+            print "Got board rev: %d" % boardRev
+        except: 
+            print "Error reading board version from i2c"
```
Comment (mjstn) [inline magni_bringup/scripts/launch_core.py]: Should we say  'Error reading motor controller board version from i2c'  here?   You decide.

File: magni_bringup/scripts/launch_core.py [hunk 2]
```diff
@@ -68,10 +70,26 @@ def get_conf():
         print "Error calling pifi"
         subprocess.call(["chronyc", "waitsync", "6"]) # Wait up to 60 seconds for chrony
 
+    boardRev = 0
```
Comment (mjstn) [inline magni_bringup/scripts/launch_core.py:73]: The firmware uses a value of 1 for unknown.   A zero would work here in firmware logic but I don't like using 0 because it is a common value that can get set on initialization and so on so 1 would be a more 'definite' value.  
We could use 1 OR could use 2 so firmware knows it was set to 2 by the host and thus the I2C was attempted to be used but failed.  This may be of some value in diagnose I2C issues.

File: magni_bringup/scripts/launch_core.py [hunk 3]
```diff
@@ -68,10 +70,26 @@ def get_conf():
...
+        except: 
+            print "Error reading board version from i2c"
+
     # Ugly, but works 
     # just passing argv doesn't work with launch arguments, so we assign sys.argv
     sys.argv = ["roslaunch", "magni_bringup", "core.launch", 
                          "raspicam_mount:=%s" % conf['raspicam']['position'],
-                         "sonars_installed:=%s" % conf['sonars']]
+                         "sonars_installed:=%s" % conf['sonars'],
+                         "controller_board_version:=%d" % boardRev]
```
Comment (mjstn) [inline magni_bringup/scripts/launch_core.py:93]: Here is where it gets tricky and we have 2 choices.   1) If a user has setup his yaml file for the version we trust that user NO MATTER WHAT and do not reset it back to something the code thinks is right.  We could tell him in a ROS error we think it should be checked/verified but we are going to 'believe' him.  2) The second more contorted method I don't like as much BUT can be argued.   In this plan IF we can read I2C AND get a firmware bitfield that is between 1 and 14 (in other words, NOT all zero or all 1) we 'trust' that value and use it for controller board version just like your logic EVEN if user has set the yaml parameter to something else.  THIS for sure we need a error at startup time to again warn we think we corrected his problem.    The tricky logic is if we cannot read a valid rev from 1-14 from I2C we only pass in controller_board_version if the user has NOT put his own controller_board_version in the yaml file.

File: magni_bringup/launch/core.launch [hunk 4]
```diff
@@ -31,13 +31,15 @@
     -->
```
Comment (mjstn) [inline magni_bringup/launch/core.launch:31]: It is important and I think it is true but must ask that  we only do this ONCE at motor node startup and we do NOT keep sending the controller board version at any other time later and certainly do not keep setting it every so often always.    True?

Comment (rohbotics) [reply magni_bringup/launch/core.launch:31]: Correct. This is only run at launch, and sets the ros parameter for the motor node. I believe the motor node only ever sends this info once.

File: magni_bringup/scripts/launch_core.py [hunk 5]
```diff
@@ -68,10 +70,26 @@ def get_conf():
...
+    boardRev = 0
+
+    if conf['motor_controller']['board_version'] == None: 
+        # Code to read board version from I2C
+        # The I2C chip is only present on 5.0 and newer boards
+        try:
+            i2cbus = smbus.SMBus(1)
+            i2cbus.write_byte(0x20, 0xFF)
+            time.sleep(0.2)
+            inputPortBits = i2cbus.read_byte(0x20)
```
Comment (mjstn) [inline magni_bringup/scripts/launch_core.py:82]: We need to test this for a couple cases and I can do this even if it duplicates your tests.
1) Obviously test with a rev 4.9 board so we can error out and not kill motor node
2) Test with no MCB board for same reason, to be sure we don't bomb if there is NOTHING o I2C at all (yes, I know we have other grief like RTC for this but still we must not make a new bug case
3) An even more obvious test is we verify on rev 5.x units that we set the rev to 50

File: magni_bringup/scripts/launch_core.py (hunk 3 above)
Comment (rohbotics) [reply magni_bringup/scripts/launch_core.py:93]: I agree with choice 1. If a user manually sets a version, they must have a deliberate and important purpose for it, so we should respect their choice. For example if they are on a non-raspberry pi platform and they don't want to write anything to that I2C port.

File: magni_bringup/scripts/launch_core.py (hunk 5 above)
Comment (rohbotics) [reply magni_bringup/scripts/launch_core.py:82]: I did tests 2 and 3, I will try out test 1.

File: magni_bringup/scripts/launch_core.py (hunk 2 above)
Comment (rohbotics) [reply magni_bringup/scripts/launch_core.py:73]: This value will not be written to the firmware unless we get a non-0 value anyway, so I don't think it is worth thinking it about too much.

File: magni_bringup/scripts/launch_core.py (hunk 3 above)
Comment (mjstn) [reply magni_bringup/scripts/launch_core.py:93]: Ok, Agreed. Choice 1.  If they set it, we respect it, whatever it may be.

File: magni_bringup/launch/core.launch (hunk 4 above)
Comment (mjstn) [reply magni_bringup/launch/core.launch:31]: Thank you, just a quick verification.

**Review by jim-v — APPROVED** [review]
Looks reasonable, but I have not tested it.

**Comment by rohbotics** [comment]
Mark, I just tried all the tests you wanted. Do you think it is good to merge now?
